src/Quadrature.h

```diff
@@ -0,0 +1,217 @@
+/**
+ * @file Quadrature.h
+ * @brief The Quadrature abstract class and subclasses.
+ * @details The Quadrature subclasses are defined with tabulated or functional
+ *          quadrature sets given in the "Lattice Physics Computations",
+ *          Handbook of Nuclear Engineering, Dave Knott, Akio Yamamoto, 2010.
+ * @date April 6, 2015
+ * @author William Boyd, MIT, Course 22 (wboyd@mit.edu)
+*/
+
+
+#ifndef QUADRATURE_H_
+#define QUADRATURE_H_
+
+#ifdef __cplusplus
+#ifdef SWIG
+#include "Python.h"
+#endif
+#include "constants.h"
+#include "log.h"
+#include <sstream>
+#endif
+
+
+/**
+ * @enum quadratureType
+ * @brief The types of quadrature sets supported by OpenMOC.
+ */
+enum quadratureType {
+  TABUCHI_YAMAMOTO,
+  LEONARD,
+  GAUSS_LEGENDRE,
+  EQUAL_WEIGHT,
+  EQUAL_ANGLE
+};
+
+
+/**
+ * @class Quadrature Quadrature.h "src/Quadrature.h"
+ * @brief The arbitrary quadrature parent class.
+ */
+class Quadrature {
+
+protected:
+
+  /** The quadrature type being used */
+  quadratureType _quad_type;
+
+  /** The number of azimuthal angles in (0, 2*PI) */
+  int _num_azim;
+
+  /** The number of polar angles in (0, PI) */
+  int _num_polar;
+
+  /** An array of the sines of quadrature polar angles */
+  FP_PRECISION** _sin_thetas;
+
+  /** An array of the quadrature polar angles */
+  double** _thetas;
+
+  /** An array of the quadrature azimuthal angles */
+  double* _phis;
+
+  /** The actual track azimuthal spacing (cm) by azimuthal angle */
+  FP_PRECISION* _azim_spacings;
+
+  /** An array of the quadrature azimuthal weights */
+  FP_PRECISION* _azim_weights;
+
+  /** An array of the quadrature polar weights */
+  FP_PRECISION** _polar_weights;
+
+  /** An array of the total weights for each azimuthal/polar angle pair */
+  FP_PRECISION** _total_weights;
+
+
+  /* Templates for setting the same values to complimentary or supplimentary
+   * angles */
+  template <typename T>
+  void setPolarValues(T** array, int azim_index, int polar_index, T value) {
```

Comment:
What is the point of using a template here in place of `FP_PRECISION`? I know this question can be flipped around on its head - e.g., "what is the point of using `FP_PRECISION` in place of fancy templating?" - but I'm not sure I see the point of templating here since there are only two real options for single vs. double precision (as far as I can tell).